### PR TITLE
Make StateController run on the same actor as SnapshotStore

### LINKED ISSUE #7889: Concurrent snapshotting and closing of the DB
**Describe the bug**

There can occur segmentation faults right now when the snapshot director requests a new snapshot be taken, and then a transition occurs which causes the `ZeebeDb` instance that's being snapshotted to be closed. This can happen concurrently since the snapshot is taken on the snapshot store actor, whereas the DB is closed in a separate actor, and the transition mechanism doesn't synchronize these operations either.

**To Reproduce**

Run the `DiskSpaceRecoveryTest` suite in a loop until failure, it usually happens quite fast on my machine.

**Expected behavior**

The access to the DB is synchronized such that once it's closed no operations can occur on it, and it cannot be closed while in operation. Doing this should prevent segmentation faults.

**Log/Stacktrace**

<details><summary>Full Stacktrace</summary>
 <p>

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007ff4f3541435, pid=6803, tid=60037
#
# JRE version: OpenJDK Runtime Environment 18.9 (11.0.10+9) (build 11.0.10+9)
# Java VM: OpenJDK 64-Bit Server VM 18.9 (11.0.10+9, mixed mode, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# C  [librocksdbjni17853063288976430087.so+0x395435]  rocksdb::DBImpl::SwitchMemtable(rocksdb::ColumnFamilyData*, rocksdb::DBImpl::WriteContext*)+0xa05
#
# Core dump will be written. Default location: /core.%e.6803.%t
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  S U M M A R Y ------------

Command Line: -XX:+UseContainerSupport -XX:MaxRAMPercentage=8 -XX:MaxDirectMemorySize=4g -Djava.net.preferIPv4Stack=true --module-path=/home/jenkins/agent/workspace/camunda-cloud_zeebe_develop/qa/integration-tests/target/zeebe-qa-integration-tests-1.2.0-SNAPSHOT.jar --patch-module=zeebe.qa.integration.tests=/home/jenkins/agent/workspace/camunda-cloud_zeebe_develop/qa/integration-tests/target/test-classes --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.client=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.client.command=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.client.multipartition=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.clustering=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.clustering.topology=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.gateway=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.health=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.network=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.smoke=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.startup=ALL-UNNAMED --add-exports=zeebe.qa.integration.tests/io.camunda.zeebe.it.system=ALL-UNNAMED --add-modules=zeebe.qa.integration.tests --add-reads=zeebe.qa.integration.tests=ALL-UNNAMED org.apache.maven.surefire.booter.ForkedBooter /home/jenkins/agent/workspace/camunda-cloud_zeebe_develop/qa/integration-tests/target/surefire 2021-09-17T16-47-58_438-jvmRun8 surefire7259606061598209383tmp surefire_1416476855135739407971tmp

Host: Intel(R) Xeon(R) CPU @ 2.30GHz, 32 cores, 48G, Debian GNU/Linux 10 (buster)
Time: Fri Sep 17 16:56:55 2021 UTC elapsed time: 261.987327 seconds (0d 0h 4m 21s)

---------------  T H R E A D  ---------------

Current thread (0x00007ff4cc09b800):  JavaThread "Broker-0-zb-fs-workers-1" daemon [_thread_in_native, id=60037, stack(0x00007ff558388000,0x00007ff558489000)]

Stack: [0x00007ff558388000,0x00007ff558489000],  sp=0x00007ff558486140,  free space=1016k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [librocksdbjni17853063288976430087.so+0x395435]  rocksdb::DBImpl::SwitchMemtable(rocksdb::ColumnFamilyData*, rocksdb::DBImpl::WriteContext*)+0xa05
C  [librocksdbjni17853063288976430087.so+0x352350]  rocksdb::DBImpl::FlushMemTable(rocksdb::ColumnFamilyData*, rocksdb::FlushOptions const&, rocksdb::FlushReason, bool)+0x3d0
C  [librocksdbjni17853063288976430087.so+0x318969]  rocksdb::DBImpl::GetLiveFiles(std::vector<std::string, std::allocator<std::string> >&, unsigned long*, bool)+0x109
C  [librocksdbjni17853063288976430087.so+0x25a5ad]  rocksdb::StackableDB::GetLiveFiles(std::vector<std::string, std::allocator<std::string> >&, unsigned long*, bool)+0x1d
C  [librocksdbjni17853063288976430087.so+0x65ea81]  rocksdb::CheckpointImpl::CreateCustomCheckpoint(rocksdb::DBOptions const&, std::function<rocksdb::Status (std::string const&, std::string const&, rocksdb::FileType)>, std::function<rocksdb::Status (std::string const&, std::string const&, unsigned long, rocksdb::FileType, std::string const&, std::string const&)>, std::function<rocksdb::Status (std::string const&, std::string const&, rocksdb::FileType)>, unsigned long*, unsigned long, bool)+0x1e1
C  [librocksdbjni17853063288976430087.so+0x6617bc]  rocksdb::CheckpointImpl::CreateCheckpoint(std::string const&, unsigned long, unsigned long*)+0x12ac
C  [librocksdbjni17853063288976430087.so+0x221df2]  Java_org_rocksdb_Checkpoint_createCheckpoint+0x62
j  org.rocksdb.Checkpoint.createCheckpoint(JLjava/lang/String;)V+0
j  org.rocksdb.Checkpoint.createCheckpoint(Ljava/lang/String;)V+6
j  io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.createSnapshot(Ljava/io/File;)V+13
j  io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl.lambda$takeSnapshot$0(Ljava/nio/file/Path;)Z+40
j  io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl$$Lambda$1831.test(Ljava/lang/Object;)Z+8
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.takeInternal(Ljava/util/function/Predicate;)V+19
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.lambda$take$0(Ljava/util/function/Predicate;)V+2
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot$$Lambda$1832.run()V+8
J 18865 c2 io.camunda.zeebe.util.sched.ActorThread.doWork()V (103 bytes) @ 0x00007ff5a8d5a018 [0x00007ff5a8d575c0+0x0000000000002a58]
J 10294 c2 io.camunda.zeebe.util.sched.ActorThread.run()V (60 bytes) @ 0x00007ff5a883aa28 [0x00007ff5a883a980+0x00000000000000a8]
v  ~StubRoutines::call_stub
V  [libjvm.so+0x8655a5]  JavaCalls::call_helper(JavaValue*, methodHandle const&, JavaCallArguments*, Thread*)+0x3e5
V  [libjvm.so+0x864983]  JavaCalls::call_virtual(JavaValue*, Klass*, Symbol*, Symbol*, JavaCallArguments*, Thread*)+0x183
V  [libjvm.so+0x864b78]  JavaCalls::call_virtual(JavaValue*, Handle, Klass*, Symbol*, Symbol*, Thread*)+0x68
V  [libjvm.so+0x90db13]  thread_entry(JavaThread*, Thread*)+0x73
V  [libjvm.so+0xd904d1]  JavaThread::thread_main_inner()+0x111
V  [libjvm.so+0xd8b18b]  Thread::call_run()+0x6b
V  [libjvm.so+0xbea198]  thread_native_entry(Thread*)+0x128

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.rocksdb.Checkpoint.createCheckpoint(JLjava/lang/String;)V+0
j  org.rocksdb.Checkpoint.createCheckpoint(Ljava/lang/String;)V+6
j  io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.createSnapshot(Ljava/io/File;)V+13
j  io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl.lambda$takeSnapshot$0(Ljava/nio/file/Path;)Z+40
j  io.camunda.zeebe.broker.system.partitions.impl.StateControllerImpl$$Lambda$1831.test(Ljava/lang/Object;)Z+8
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.takeInternal(Ljava/util/function/Predicate;)V+19
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot.lambda$take$0(Ljava/util/function/Predicate;)V+2
j  io.camunda.zeebe.snapshots.impl.FileBasedTransientSnapshot$$Lambda$1832.run()V+8
J 18865 c2 io.camunda.zeebe.util.sched.ActorThread.doWork()V (103 bytes) @ 0x00007ff5a8d5a018 [0x00007ff5a8d575c0+0x0000000000002a58]
J 10294 c2 io.camunda.zeebe.util.sched.ActorThread.run()V (60 bytes) @ 0x00007ff5a883aa28 [0x00007ff5a883a980+0x00000000000000a8]
v  ~StubRoutines::call_stub

siginfo: si_signo: 11 (SIGSEGV), si_code: 1 (SEGV_MAPERR), si_addr: 0x0000000000000040

```

</p>
</details>

**Environment:**
 - OS: 
 - Zeebe Version: 
 - Configuration:

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

To prevent the race condition between closing the database and taking snapshot, we now run all operations on StateController on the same actor as SnapshotStore.

* Changed StateController interface to asynchronous, all operations return future.
* Removed some methods from StateController to simplify them
* StateControllerImpl now takes a "ConcurrencyControl" instance and use that to execute all operation. The ConcurrencyControl is set to the SnapshotStore actor

## Related issues

closes #7889 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java [hunk 1]
```diff
@@ -347,7 +347,7 @@ public FileBasedReceivedSnapshot newReceivedSnapshot(final String snapshotId) {
   }
 
   private void addPendingSnapshot(final PersistableSnapshot pendingSnapshot) {
-    actor.call(() -> pendingSnapshots.add(pendingSnapshot));
+    actor.run(() -> pendingSnapshots.add(pendingSnapshot));
```
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:350]: :x: You can use submit now even from outside the actor, and it guarantees you get append semantics (and not prepend when you call it from an actor thread, as with run). The one caveat is that submitted tasks via submit will not be ran during the closing phase, whereas those submitted with `run` will. But that was already the case with `call`, so replacing `call` with `submit` is safe.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStoreFactory.java [hunk 2]
```diff
@@ -86,4 +87,15 @@ public ReceivableSnapshotStore getReceivableSnapshotStore(final int partitionId)
...
     return partitionSnapshotStores.get(partitionId);
   }
+
+  /**
+   * Return the same concurrent control (actor) that is used by the snapshot store of the given
+   * partition
+   *
+   * @param partitionId
+   * @return concurrency control
+   */
```
Comment (npepinpe) [inline snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStoreFactory.java:97]: :wrench: I would mark it as deprecated already to discourage usage, and we should work on refactoring this soon-ish

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

Generally looks good, and I agree that we should refactor very soon to introduce the concept state actor/context which manages both the snapshot store and the state controller. It might be easier to do this though after we've deconstructed the Atomix monolith, i.e. when the startup phase will build the snapshot store as a step, pass it to the Raft component and so on, and that we don't have factories anymore.

Anyway, some notes:

- :x: Can we have a test which makes sure you cannot close the DB concurrently when you're taking a snapshot of it? I'm sorry if I missed it and it's there, then just point it out to me.

File: snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java (hunk 1 above)
Comment (deepthidevaki) [reply snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedSnapshotStore.java:350]: Thanks. Didn't realize "submit" has a new semantics.

**Comment by deepthidevaki** [comment]
> Can we have a test which makes sure you cannot close the DB concurrently when you're taking a snapshot of it? I'm sorry if I missed it and it's there, then just point it out to me.

I don't know how to reliably test the concurrency part. I wrote a test, but the test didn't fail with out this fix.

**Comment by npepinpe** [comment]
It would be easier if the `FileBasedSnapshotStore` was not itself an actor, but simply received a concurrency control :smile: 

At any rate, it should be possible with a `ControllableActorScheduler`, no? There you control when jobs are executed, IIRC. So you could pause it, enqueue the jobs, close the DB, let it work again?

Ah, or was this error _only_ happening when you close while it's already told RocksDB to take the snapshot (i.e. the RocksDB code is being executed while we close it)? Then yeah, I'm not sure how we would test the exact bug.

But we can still write a defensive test, no? We can write a test which makes sure taking a snapshot and closing cannot overlap?

**Comment by deepthidevaki** [comment]
> Ah, or was this error only happening when you close while it's already told RocksDB to take the snapshot (i.e. the RocksDB code is being executed while we close it)? Then yeah, I'm not sure how we would test the exact bug.

Ya. Something like that.
On the takeSnapshot() we have following code:
```
if(!dbIsOpen) {
   return
} else {
   take()
}
```
So previously, the verification `dbIsOpen` passes( on snapshot actor), then db is closed (on the actor), then snapshot is taken (snapshot actor), and this fails.

**Comment by deepthidevaki** [comment]
> At any rate, it should be possible with a ControllableActorScheduler, no? There you control when jobs are executed, IIRC. So you could pause it, enqueue the jobs, close the DB, let it work again?

Where do we add this test? In the unit tests for StateControllerImpl? There we are already scheduling both state controller and snapshot store on the same actor. So we can rely on ActorScheduler to serialize the jobs. So does it make sense to test the concurrency, as there is no concurrency?

**Comment by npepinpe** [comment]
That's more of an integration test with the transitions, isn't it? Do we have any of these to verify the impact of transitions on the partition state as a whole and other components? 

That said, I just tried writing the pseudo code for it, and unfortunately I don't think it'll be possible. Since what you want to test is the absence of something - i.e. when I'm in the middle of taking a snapshot, and a transition from LEADER to FOLLOWER occurs, then the transition is blocked until I take a snapshot. But how would you verify this? A time out? :smile: That sounds like it's going to be flaky. I can't think of something else, so I guess unit tests will have to do...but I wish we could have something to guarantee this behavior =/

**Comment by deepthidevaki** [comment]
I added simple unit tests for it. No concurrency. Please take a look if that is enough https://github.com/camunda-cloud/zeebe/pull/7898/commits/6439b8c165e623ebfbf2db1f46cd19b0768c1239

**Review by npepinpe — APPROVED** [review]
:+1: 

Thanks!

**Comment by deepthidevaki** [comment]
Flaky shouldExportEvenIfElasticNotInitiallyReachable (Flaky Test) – ElasticsearchExporterFaultToleranceIT(it-testrun)

bors retry

**Pull request merged**